src/main/kotlin/com/redhat/devtools/intellij/kubernetes/model/context/LazyOpenShiftContext.kt

```diff
@@ -0,0 +1,112 @@
+import com.redhat.devtools.intellij.kubernetes.model.resource.IResourceOperator
+import com.redhat.devtools.intellij.kubernetes.model.resource.ResourceKind
+import com.redhat.devtools.intellij.kubernetes.model.resource.kubernetes.KubernetesReplicas.Replicator
+import io.fabric8.kubernetes.api.model.HasMetadata
+import io.fabric8.kubernetes.api.model.NamedContext
+import io.fabric8.kubernetes.client.KubernetesClient
+import org.jetbrains.concurrency.runAsync
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.write
+
+/**
+ * A delegating context that starts with a Kubernetes delegate.
+ * It then (async) tries to create and use an OpenShift context, notifying the change if it successfully could.
+ * It sticks to the Kubernetes if it can't.
+ *
+ * @see createOpenShiftDelegate
+ */
+class LazyOpenShiftContext(
+    context: NamedContext,
+    modelChange: IResourceModelObservable,
+    client: KubeClientAdapter,
+    /* for testing purposes */
+    kubernetesContextFactory: (
+        context: NamedContext,
+        modelChange: IResourceModelObservable,
+        client: KubeClientAdapter,
+    ) -> IActiveContext<out HasMetadata, out KubernetesClient>
+    = ::KubernetesContext,
+    /* for testing purposes */
+    private val openshiftContextFactory: (
+        client: OSClientAdapter,
+        modelChange: IResourceModelObservable
+    ) -> IActiveContext<out HasMetadata, out KubernetesClient>?
+    = IActiveContext.Factory::createOpenShift,
+    /* for testing purposes */
+    runAsync: (runnable: () -> Unit) -> Unit
+    = ::runAsync
+) : KubernetesContext(context, modelChange, client) {
+
+    private val lock = ReentrantReadWriteLock()
+    private var delegate: IActiveContext<out HasMetadata, out KubernetesClient> =
+        kubernetesContextFactory.invoke(context, modelChange, client)
+
+    init {
+        runAsync.invoke {
+            createOpenShiftDelegate()
+        }
+    }
+
+    override val namespaceKind: ResourceKind<out HasMetadata>
+        get() = delegate.namespaceKind
+
+    override fun getInternalResourceOperators(): List<IResourceOperator<out HasMetadata>> {
+        lock.read {
+            return delegate.getInternalResourceOperators()
+        }
+    }
+
+    override fun isOpenShift(): Boolean {
+        lock.read {
+            return delegate.isOpenShift()
+        }
+    }
+
+    override fun setReplicas(replicas: Int, replicator: Replicator) {
+        lock.read {
+            delegate.setReplicas(replicas, replicator)
+        }
+    }
+
+    override fun getReplicas(resource: HasMetadata): Replicator? {
+        lock.read {
+            return delegate.getReplicas(resource)
+        }
+    }
+
+    override fun getDashboardUrl(): String? {
+        lock.read {
+            return delegate.getDashboardUrl()
+        }
+    }
+
+    private fun createOpenShiftDelegate() {
+        if (client.canAdaptToOpenShift()) {
```

Comment:
this is the core fix: a context that starts as a Kubernetes context and async tries to create and use an OpenShift context.